srml/authority-discovery/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Authority discovery module.
+//!
+//! This module is used by the `core/authority-discovery` to retrieve the
+//! current set of authorities, learn its own authority id as well as sign and
+//! verify messages to and from other authorities.
+//!
+//! ## Dependencies
+//!
+//! This module depends on the [Im online module](../srml_im_online/index.html)
+//! using its session key.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use app_crypto::RuntimeAppPublic;
+use codec::{Decode, Encode};
+use rstd::prelude::*;
+use srml_support::{decl_module, decl_storage, StorageValue};
+
+pub trait Trait: system::Trait + session::Trait {}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as AuthorityDiscovery {
+		/// The current set of keys that may issue a heartbeat.
+		Keys get(keys): Vec<im_online::AuthorityId>;
+	}
+	add_extra_genesis {
+		config(keys): Vec<im_online::AuthorityId>;
+		build(|
+			  storage: &mut (sr_primitives::StorageOverlay, sr_primitives::ChildrenStorageOverlay),
+			  config: &GenesisConfig
+			  | {
+				  sr_io::with_storage(
+					  storage,
+					  || Module::<T>::initialize_keys(&config.keys),
+				  );
+			  })
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Returns own authority identifier iff it is part of the current authority
+	/// set, otherwise this function returns None. The restriction might be
+	/// softened in the future in case a consumer needs to learn own authority
+	/// identifier in any case.
+	pub fn public_key() -> Option<im_online::AuthorityId> {
```

Comment:
This is not scalable (`O(n)`) if we have a large authority set.

Reply:
@DemiMarie-parity can you add some details to your comment? Are you referring to generating the intersection of two authority id sets?

I have written a small benchmark with the assumption of 1000 current authorities (*n*) and 10 local authority keys (*m*). In the worst case we have to iterate the entire set *n* and compare it to each key in *m* resulting in an `O(n*m)`. Running this worst case scenario on a fairly new laptop compiled to machine code (not web assembly) takes ~13 us.

Is that within a range that we should worry about?

Source code can be found here: https://github.com/mxinden/benchmark-authority-id-intersection/blob/master/benches/intersection.rs#L13

Reply:
@mxinden Yes, I am.  My worry is that this could wind up being called by wasm code in a tight loop, which would be slow.

Reply:
The code is not running on-chain. So, even if we iterate 1mio times, it should not be any bottleneck.